waveorder/stokes.py

```diff
@@ -0,0 +1,344 @@
+    range convention used in comp-micro: 0 <= ori < pi.
+
+    Parameters
+    ----------
+    s1, s2: NDArray, identical shapes
+        Stokes parameters
+
+    Returns
+    -------
+    NDArray
+        Slow-axis orientation with 0 <= ori < pi.
+    """
+    return (np.arctan2(s1, -s2) % (2 * np.pi)) / 2
+
+
+def inverse_s0123_CPL_after_ADR(s0, s1, s2, s3):
+    """
+    Inverse of s0123_CPL_after_ADR.
+
+    Given the Stokes parameters of circularly polarized light (CPL) that has
+    passed through an attenuating depolarizing retarder (ADR), this function
+    returns the parameters of the ADR, specifically its retardance (ret),
+    slow-axis orientation (ori), transmittance (tra), and depolarization (dop).
+
+    Note: this function is commonly used in QLIPP-type reconstructions. After
+    converting raw intensities into Stokes parameters by applying A_inv, the
+    Stokes parameters are uses to estimate the parameters of the sample ADR in
+    a single step with this function.
+
+    Parameters
+    ----------
+    s0, s1, s2, s3: NDArray, identical shapes
+        Stokes parameters
+
+    Returns
+    ----------
+    ret, ori, tra, dop: NDArray
+        ret: retardance of ADR, 2*pi periodic
+        ori: slow-axis orientation of ADR, 2*pi periodic
+        tra: transmittance of ADR, 0 <= tra <= 1
+        dop: depolarization of ADR, 0 <= dop <= 1
+    """
+    len_pol = (s1**2 + s2**2 + s3**2) ** 0.5
+    ret = np.arcsin(((s1**2 + s2**2) ** 0.5) / len_pol)
+    ori = _s12_to_ori(s1, s2)
+    tra = s0
+    dop = len_pol / s0
+    return ret, ori, tra, dop
+
+
+def inverse_s012_CPL_after_AR(s0, s1, s2):
+    """
+    Inverse of s012_CPL_after_ADR.
+
+    Given the Stokes parameters of circularly polarized light (CPL) that has
+    passed through an attenuating retarder (AR), this function returns the
+    parameters of the ADR, specifically its retardance (ret), slow-axis
+    orientation (ori), and transmittance (tra).
+
+    Parameters
+    ----------
+    s0, s1, s2: NDArray, identical shapes
+        First three Stokes parameters
+
+    Returns
+    ----------
+    ret, ori, tra: NDArray, identical shapes
+        ret: retardance of ADR, 2*pi periodic
+        ori: slow-axis orientation of ADR, 2*pi periodic
+        tra: transmittance of ADR, 0 <= tra <= 1
+    """
+    ret = np.arcsin(((s1**2 + s2**2) ** 0.5) / s0)
+    ori = _s12_to_ori(s1, s2)
+    tra = s0
+    return ret, ori, tra
+
+
+def AR_mueller_from_CPL_projection(s0, s1, s2, s3):
+    """
+    Given the Stokes parameters of circularly polarized light (CPL) that has
+    passed through an attenuating retarder (AR), this function returns the
+    complete Mueller matrix of the AR.
+
+    Parameters
+    ----------
+    s0, s1, s2, s3 : NDArray, identical shapes
+        Stokes parameters
+
+    Returns
+    -------
+    NDArray, float, M.shape = (4, 4,) + s0.shape
+        Mueller matrix
+    """
+    M = np.zeros((4, 4) + np.array(s0).shape)
```

Comment:
Why is `np.array()` needed if `s0` is already an `NDArray` object?

Reply:
Because it's convenient to be able to provide an `NDArray` or a raw `float` (as a proxy for a 0D `NDArray`). 

I'm open to pushback if this is a bad idea.

Reply:
Then maybe annotate here to?
https://github.com/mehta-lab/waveorder/blob/56eb82744e7c0cde712419b9cc4d3bae45cbbc83/waveorder/stokes.py#L268-L269

Reply:
I'm considering changing my docs from `NDArray` back to `array_like` since that's what `numpy` does.

I'm aiming for behavior like most numpy functions. If you pass in "raw numbers", it'll return a raw number (`type(np.arctan2(1.0,1.0))` gives `np.float64`), and if you pass in an array it'll return an array (`type(np.arctan2(np.array([1.0,2.0]),np.array([1.0,2.0])))` gives `np.ndarray`).

Reply:
I think documenting with the `array_like` type implies these cases, since it's how `numpy` is documented.

Reply:
`ArrayLike` is [a `Union` of many different things](https://github.com/numpy/numpy/blob/main/numpy/_typing/_array_like.py). If the purpose is to accept `float`s and `NDArray`s, but not Dask arrays (implements array protocol) and lists of lists (nested sequence), `Union[np.floating, NDArray]` is more accurate.

Reply:
These functions are thin wrappers for `numpy` functions that are documented as accepting `array_like` types, so I don't understand why I can't also document these functions as accepting `array_like` types. They should work on Dask arrays and lists of lists if the underlying `numpy` arrays also accept these. I'm only using `.shape` and `np` functions that accept `array_like`s.